src/Raven.Server/Documents/PeriodicBackup/BackupTask.cs

```diff
@@ -659,6 +676,78 @@ private InternalBackupResult CreateLocalBackupOrSnapshot(PeriodicBackupStatus st
+        {
+            switch (_directUploadDestination)
+            {
+                case DirectUploadDestination.Disabled:
+                    return SafeFileStream.Create(filePath, FileMode.Create);
+
+                    case DirectUploadDestination.S3:
+                        var s3Settings = GetBackupConfigurationFromScript(_configuration.S3Settings, x => JsonDeserializationServer.S3Settings(x),
+                            settings => PutServerWideBackupConfigurationCommand.UpdateSettingsForS3(settings, _database.Name));
+
+                        return new AwsS3DirectUploadStream(new DirectUploadStream<RavenAwsS3Client>.Parameters
+                        {
+                            Client = progress => new RavenAwsS3Client(s3Settings, _database.Configuration.Backup, progress, TaskCancelToken.Token),
+                            Key = BackupUploader.CombinePathAndKey(s3Settings.RemoteFolderName, folderName, fileName),
+                            Metadata = new Dictionary<string, string>
+                            {
+                                { "Description", BackupUploader.GetBackupDescription(_configuration.BackupType, _isFullBackup) }
+                            },
+                            IsFullBackup = _isFullBackup,
+                            RetentionPolicyParameters = _retentionPolicyParameters,
+                            CloudUploadStatus = _backupResult.S3Backup,
+                            OnProgress = AddInfo,
+                        });
+
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+        }
+
+        private DirectUploadDestination GetDirectUploadDestination()
+        {
+            if (_backupToLocalFolder)
+            {
+                // we'll do the local backup and then upload it
+                return DirectUploadDestination.Disabled;
+            }
+
+            if (_database.Configuration.Backup.DisableDirectUpload)
+            {
+                // disabled by configuration
+                return DirectUploadDestination.Disabled;
+            }
+
+            var hasAws = BackupConfiguration.CanBackupUsing(_configuration.S3Settings);
+            var hasGlacier = BackupConfiguration.CanBackupUsing(_configuration.GlacierSettings);
+            var hasAzure = BackupConfiguration.CanBackupUsing(_configuration.GlacierSettings);
+            var hasGoogleCloud = BackupConfiguration.CanBackupUsing(_configuration.GoogleCloudSettings);
+            var hasFtp = BackupConfiguration.CanBackupUsing(_configuration.GoogleCloudSettings);
+
+            var destinations = new List<bool> { hasAws, hasGlacier, hasAzure, hasGoogleCloud, hasFtp };
+            if (destinations.Count(x => x) != 1)
+            {
+                // direct upload is supported only for 1 destination
+                return DirectUploadDestination.Disabled;
+            }
```

Comment:
why do you need that?

Reply:
The optimization is only when you have a single non-local destination.
For example, if you have a backup to `Azure` and `AWS` this will require us to read the same data file twice (if using the direct upload).

Reply:
okay, later this should be explain in the docs/studio